File: Day8.py
data = [(i.split()[0], int(i.split()[1])) for i in open('./input.txt').read().splitlines()]

def ends(data, part1 = False):
    visited = [False]*len(data)
    total, index = 0, 0
    while True:
        if index >= len(data):
            break
        if visited[index]:
            return total if part1 else False
        else:
            visited[index] = True
        if data[index][0] == 'acc': 
            total += data[index][1]
        if data[index][0] == 'jmp':
            index += data[index][1]
            continue
        index += 1
    return total


print('Part1: ' + str(ends(data, part1=True)))

for i,d in enumerate(data):
    if d[0] == 'acc': continue
    data[i] = ('jmp', d[1]) if d[0] == 'nop' else ('nop', d[1])
    total =  ends(data)
    if total:
        break
    data[i] = d # restore the original
    
print(total)

File: Day8.js
// AoC 2020 Day 8
// 12/8/2020 1:00:06 AM

const fs = require('fs');

p = console.log

var path = 'input.txt'

var text = fs.readFileSync(path,'utf8').split(/\r?\n/)
const data = text.map(x => x.split(' '));

const ends = (data, part1= false) => {
    let visited = Array(data.length).fill(false);
    let total = 0;
    let index = 0;
    while(true){
        if(index >= data.length) break;
        if(visited[index]) 
            return part1 ? total : false
        else
            visited[index] = true
        if(data[index][0] === 'acc')
            total += parseInt(data[index][1])
        if(data[index][0] === 'jmp'){
            index += parseInt(data[index][1])
            continue
        }
        index++;
    }
    return total;
}

p('Part1: ' + ends(data, true))

for(const [i, d] of data.entries()){
    if(d[0] === 'acc') continue;
    data[i] = (d[0] === 'nop') ? ['jmp', d[1]] : ['nop', d[1]]
    var total = ends(data)
    if(total) break;
    data[i] = d
}
p('Part2: ' + total)